iceaxe/__tests__/conftest.py

```diff
@@ -1,84 +1,300 @@
+        logger.info(f"Using preferred port {port}")
+    else:
+        port = get_free_port()
+        logger.info(f"Using alternative port {port}")
+
+    # PostgreSQL connection details
+    pg_user = "iceaxe"
+    pg_password = "mysecretpassword"
+    pg_db = "iceaxe_test_db"
+
+    # Start PostgreSQL container
+    try:
+        container = client.containers.run(
+            "postgres:16",
+            name=container_name,
+            detach=True,
+            environment={
+                "POSTGRES_USER": pg_user,
+                "POSTGRES_PASSWORD": pg_password,
+                "POSTGRES_DB": pg_db,
+            },
+            ports={"5432/tcp": port},
+            remove=True,  # Auto-remove container when stopped
+        )
+    except APIError as e:
+        # If there's still an issue, try with a random port as a last resort
+        if "port is already allocated" in str(e):
+            logger.warning(
+                f"Port {port} is still in use. Trying with a completely random port."
+            )
+            port = get_free_port()
+            try:
+                container = client.containers.run(
+                    "postgres:16",
+                    name=container_name,
+                    detach=True,
+                    environment={
+                        "POSTGRES_USER": pg_user,
+                        "POSTGRES_PASSWORD": pg_password,
+                        "POSTGRES_DB": pg_db,
+                    },
+                    ports={"5432/tcp": port},
+                    remove=True,
+                )
+            except Exception as inner_e:
+                raise RuntimeError(
+                    f"Failed to start PostgreSQL container with random port: {inner_e}"
+                ) from e
+        else:
+            raise RuntimeError(f"Failed to start PostgreSQL container: {e}")
+
+    # Wait for PostgreSQL to be ready
+    max_retries = 30
+    retry_interval = 1
+    for i in range(max_retries):
+        container.reload()  # Refresh container status
+        if container.status != "running":
+            raise RuntimeError(f"Container failed to start: {container.status}")
+
+        # Try to connect to PostgreSQL
+        try:
+            conn = socket.create_connection(("localhost", port), timeout=1)
+            conn.close()
+            break
+        except (socket.error, ConnectionRefusedError):
+            if i == max_retries - 1:
+                container.stop()
+                raise RuntimeError("Failed to connect to PostgreSQL container")
+            time.sleep(retry_interval)
+
+    # Wait a bit more to ensure PostgreSQL is fully initialized
+    time.sleep(2)
+
+    # Yield connection details
+    connection_info = {
+        "host": "localhost",
+        "port": port,
+        "user": pg_user,
+        "password": pg_password,
+        "database": pg_db,
+    }
+
+    yield connection_info
+
+    # Cleanup: stop the container
+    try:
+        container.stop()
```

Comment:
I sometimes find myself wanting to analyze the state of the database after tests have completed with `psql`. Would it make sense to add a `finally` block here that reads an env variable `ICEAXE_TESTING_CLEAN_DB=FALSE` that will keep it open for X minutes after the yield instead of immediately closing?

Reply:
That is definitely an option, but it would probably be easier to set a breakpoint after a given test. Let me know where you want me to do.